File: Source/Global/Globals.cs
using System;
using System.Collections.Generic;
using WCSharp.Api;
using static WCSharp.Api.Common;

public static class Globals
{
    public const int MAX_TEAM_SIZE = 12;
    public const float TIME_TO_PICK_GAMEMODE = 20.0f;
    public const int DEFAULT_TEAM_SIZE = 3;
    public static int ROUND = 0;
    public static bool GAME_ACTIVE { get; set; } = false;
    public static float GAME_SECONDS = 0.0f;
    public static rect WORLD_BOUNDS = rect.CreateWorldBounds();
    public static int GAME_SEED { get; set; }
    public static List<Kitty> TempKittyList { get; } = new List<Kitty>();
    public static readonly GameMode[] GAME_MODES = { GameMode.Standard, GameMode.Solo, GameMode.Team };
    public static readonly string[] TEAM_MODES = { "Free Pick", "Random" };
    public static readonly string[] SOLO_MODES = { "Progression", "Race" };
    public static readonly string[] DeveloperList = { "QWNoZXMjMTgxNw==", "TG9jYWwgUGxheWVy", "U3RhbiMyMjM5OQ==", "WW9zaGltYXJ1IzIxOTc2" };
    public static readonly string[] AdminList = { "Q2FpdCMxMjgwNQ==", "T21uaW9sb2d5IzExODUw", "Rmllcnlmb3gjMjE2NDA=", "S2Fub24yODkjMTU5Mw==" };
    public static readonly string[] VipList = { "Q2FydG1hbiMyMzMxNQ==", "S3ltcCMyNDUw", "TXJHaGVlZCMxODMx",
        "aG9mZiMxMTQwNA==", "SmFtZXNGcmFuY28jMTE3MTk=", "TmF0aGFuc2VycG8jMjQ2MQ==",
        "QmFsbXlkcm9wIzE3Mzc=", "QnJhbkZsYWtlNjQjMTEyNw==", "Tm9vYmVybWFuIzExNTc5", "TmF6ZSMxMTI2OQ==", "T3JiaXQjMTc1MQ==" };
    public static List<player> DEVELOPER_LIST = new List<player>();
    public static List<player> ADMIN_LIST = new List<player>();
    public static List<player> VIP_LIST = new List<player>();

    public static readonly string[] CHAMPIONS =
        { "Aches#1817", "Fieryfox#21640", "Qoz#11803", "BranFlake64#1127", "BranFlake#1127",
        "Balmydrop#1737", "udo#11673", "MrGheed#1831", "Local Player", "Stan#22399",
        "Omniology#11850", "Danger#24279", "nomordarkwar#2525"};

    public static timer GAME_TIMER = timer.Create();
    public static timerdialog GAME_TIMER_DIALOG = CreateTimerDialog(GAME_TIMER);

    public static List<player> ALL_PLAYERS { get; set; } = new List<player>();
    public static List<Safezone> SAFE_ZONES { get; set; } = new List<Safezone>();
    public static List<player> LockedCamera { get; set; } = new List<player>();
    public static List<Kitty> ALL_KITTIES_LIST { get; set; } = new List<Kitty>();


    public static Dictionary<player, Kitty> ALL_KITTIES { get; set; } = new Dictionary<player, Kitty>();
    public static Dictionary<player, Circle> ALL_CIRCLES { get; set; } = new Dictionary<player, Circle>();
    public static Dictionary<unit, Wolf> ALL_WOLVES { get; set; } = new Dictionary<unit, Wolf>();
    public static List<Wolf> ALL_WOLVES_LIST { get; set; } = new List<Wolf>();

    public static Dictionary<player, PlayerUpgrades> PLAYER_UPGRADES { get; set; } = new Dictionary<player, PlayerUpgrades>();

    public static SaveManager SaveSystem { get; set; }
    public static GameAwardsDataSorted GAME_AWARDS_SORTED { get; } = new GameAwardsDataSorted();
    public static RoundTimesData GAME_TIMES { get; set; } = new RoundTimesData();
    public static GameStatsData GAME_STATS { get; set; } = new GameStatsData();
    // public static GameTimesData SAVE_GAME_ROUND_DATA = new GameTimesData();

    /// <summary>
    /// Seeded Random GEN for things that need to be based on the game seed, such as affixes or save system reasons.
    /// </summary>
    public static Random RANDOM_GEN { get; set; }
    /// <summary>
    /// Unseeded Random GEN for things that need to be more random and not based on the game seed. (Tournament saver)
    /// </summary>
    public static Random RANDOM_GEN_02 { get; set; }
    public static bool DATE_TIME_LOADED { get; set; }
    public static bool GAME_INITIALIZED { get; set; }

}

File: src/Global/Globals.ts
import { Circle } from 'src/Game/Entities/Circle'
import { Kitty } from 'src/Game/Entities/Kitty/Kitty'
import { Wolf } from 'src/Game/Entities/Wolf'
import { PlayerUpgrades } from 'src/Game/Items/Relics/PlayerUpgrades'
import { Safezone } from 'src/Game/Management/Safezone'
import { Team } from 'src/Gamemodes/Teams/Team'
import { GameAwardsDataSorted } from 'src/SaveSystem2.0/MAKE REWARDS HERE/SaveObjects/GameAwardsDataSorted'
import { GameStatsData } from 'src/SaveSystem2.0/MAKE REWARDS HERE/SaveObjects/GameStatsData'
import { RoundTimesData } from 'src/SaveSystem2.0/MAKE REWARDS HERE/SaveObjects/RoundTimesData'
import { SaveManager } from 'src/SaveSystem2.0/SaveManager'
import { Group, MapPlayer, Rectangle, Timer, TimerDialog, Unit } from 'w3ts'
import { GameMode } from '../Gamemodes/GameModeEnum'

export class Globals {
    public static MAX_TEAM_SIZE: number = 24
    public static DEFAULT_TEAM_SIZE: number = 3
    public static TIME_TO_PICK_GAMEMODE: number = 20.0
    public static ROUND: number = 0
    public static GAME_ACTIVE: boolean = false
    public static GAME_SECONDS: number = 0.0
    public static WORLD_BOUNDS: Rectangle = Rectangle.getWorldBounds()!
    public static GAME_SEED: number
    public static TempGroup: Group = Group.create()!
    public static readonly GAME_MODES: string[] = [
        GameMode[GameMode.Standard],
        GameMode[GameMode.SoloTournament],
        GameMode[GameMode.TeamTournament],
    ]
    public static readonly TEAM_MODES: string[] = ['Pick: Free', 'Random']
    public static readonly SOLO_MODES: string[] = ['Progression', 'Race']
    public static readonly VIPLIST: string[] = [
        'QWNoZXMjMTgxNw==',
        'TG9jYWwgUGxheWVy',
        'Q2FpdCMxMjgwNQ==',
        'T21uaW9sb2d5IzExODUw',
        'U3RhbiMyMjM5OQ==',
        'WW9zaGltYXJ1IzIxOTc2',
    ]
    public static VIPLISTUNFILTERED: MapPlayer[] = []

    public static readonly CHAMPIONS: string[] = [
        'Aches#1817',
        'Fieryfox#21640',
        'Qoz#11803',
        'BranFlake64#1127',
        'BranFlake#1127',
        'Balmydrop#1737',
        'udo#11673',
        'MrGheed#1831',
        'Player: Local',
        'Stan#22399',
        'Omniology#11850',
        'Danger#24279',
    ]

    public static GAME_TIMER = Timer.create()
    public static GAME_TIMER_DIALOG = TimerDialog.create(Globals.GAME_TIMER)!

    public static ALL_PLAYERS: MapPlayer[] = []
    public static SAFE_ZONES: Safezone[] = []
    public static LockedCamera: MapPlayer[] = []
    public static ALL_KITTIES_LIST: Kitty[] = []

    public static ALL_KITTIES: Map<MapPlayer, Kitty> = new Map()
    public static ALL_CIRCLES: Map<MapPlayer, Circle> = new Map()
    public static ALL_WOLVES: Map<Unit, Wolf> = new Map()

    public static PLAYER_UPGRADES: Map<MapPlayer, PlayerUpgrades> = new Map()

    public static SaveSystem: SaveManager
    public static GAME_AWARDS_SORTED: GameAwardsDataSorted = new GameAwardsDataSorted()
    public static GAME_TIMES: RoundTimesData = new RoundTimesData()
    public static GAME_STATS: GameStatsData = new GameStatsData()
    // public static GameTimesData SAVE_GAME_ROUND_DATA = new GameTimesData();

    public static ALL_TEAMS: Map<number, Team> = new Map()

    public static ALL_TEAMS_LIST: Team[] = []
    public static PLAYERS_TEAMS: Map<MapPlayer, Team> = new Map()
    public static TEAM_PROGRESS: Map<Team, string> = new Map()

    public static DATE_TIME_LOADED: boolean
    public static GAME_INITIALIZED: boolean

    // Round                         // Lane ,  // Updated # Wolves (+5% for all lanes, first 6 + 5%)
    public static WolvesPerRound: Map<number, { [lane: number]: number }> = new Map([
        [
            1,
            {
                0: 25,
                1: 25,
                2: 25,
                3: 25,
                4: 17,
                5: 16,
                6: 16,
                7: 15,
                8: 12,
                9: 12,
                10: 12,
                11: 7,
                12: 7,
                13: 6,
                14: 6,
                15: 2,
                16: 2,
            },
        ],
        [
            2,
            {
                0: 34,
                1: 34,
                2: 34,
                3: 34,
                4: 23,
                5: 22,
                6: 20,
                7: 19,
                8: 15,
                9: 15,
                10: 14,
                11: 10,
                12: 10,
                13: 8,
                14: 7,
                15: 3,
                16: 3,
            },
        ],
        [
            3,
            {
                0: 39,
                1: 39,
                2: 38,
                3: 38,
                4: 27,
                5: 26,
                6: 24,
                7: 22,
                8: 17,
                9: 17,
                10: 16,
                11: 12,
                12: 12,
                13: 8,
                14: 8,
                15: 3,
                16: 3,
            },
        ],
        [
            4,
            {
                0: 48,
                1: 48,
                2: 48,
                3: 46,
                4: 32,
                5: 31,
                6: 27,
                7: 26,
                8: 20,
                9: 20,
                10: 18,
                11: 14,
                12: 14,
                13: 10,
                14: 9,
                15: 4,
                16: 4,
            },
        ],
        [
            5,
            {
                0: 57,
                1: 57,
                2: 57,
                3: 57,
                4: 37,
                5: 35,
                6: 30,
                7: 29,
                8: 21,
                9: 21,
                10: 20,
                11: 16,
                12: 16,
                13: 11,
                14: 10,
                15: 5,
                16: 5,
            },
        ],
    ])

    /// <summary>
    /// Mock Data to Test Whatever lanes, and # of wolves as needed.
    /// </summary>
    // public static WolvesPerRound: Map<number, { [lane: number]: number> } = {
    //     // Round                         // Lane ,  // Updated # Wolves (+5% for all lanes, first 6 + 5%)
    //     1: { 0: 25, 1: 25, 2: 25, 3: 25, 4: 17, 5: 16, 6: 16, 7: 15, 8: 12, 9: 12, 10: 12, 11: 7, 12: 7, 13: 6, 14: 6, 15: 2, 16: 2 },
    //     2: { 0: 34, 1: 34, 2: 34, 3: 34, 4: 23, 5: 22, 6: 20, 7: 19, 8: 15, 9: 15, 10: 14, 11: 10, 12: 10, 13: 8, 14: 7, 15: 3, 16: 3 },
    //     3: { 0: 39, 1: 39, 2: 38, 3: 38, 4: 27, 5: 26, 6: 24, 7: 22, 8: 17, 9: 17, 10: 16, 11: 12, 12: 12, 13: 8, 14: 8, 15: 3, 16: 3 },
    //     4: { 0: 48, 1: 48, 2: 48, 3: 46, 4: 32, 5: 31, 6: 27, 7: 26, 8: 20, 9: 20, 10: 18, 11: 14, 12: 14, 13: 10, 14: 9, 15: 4, 16: 4 },
    //     5: { 0: 57, 1: 57, 2: 57, 3: 57, 4: 37, 5: 35, 6: 30, 7: 29, 8: 21, 9: 21, 10: 20, 11: 16, 12: 16, 13: 11, 14: 10, 15: 5, 16: 5 }
    // };
}
